pkg/output/json.go

```diff
@@ -0,0 +1,147 @@
+/*
+==========
+Cariddi
+==========
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see http://www.gnu.org/licenses/.
+
+	@Repository:  https://github.com/edoardottt/cariddi
+
+	@Author:      edoardottt, https://www.edoardoottavianelli.it
+
+	@License: https://github.com/edoardottt/cariddi/blob/main/LICENSE
+
+*/
+
+package output
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+
+	"github.com/edoardottt/cariddi/pkg/scanner"
+	"github.com/gocolly/colly"
+)
+
+type JSONData struct {
+	URL           string         `json:"url"`
+	Method        string         `json:"method"`
+	StatusCode    int            `json:"status_code"`
+	Words         int            `json:"words"`
+	Lines         int            `json:"lines"`
+	ContentType   string         `json:"content_type,omitempty"`
+	ContentLength int            `json:"content_length,omitempty"`
+	Matches       MatcherResults `json:"matches,omitempty"`
+	// Host          string `json:"host"` # TODO: Available in Colly 2.x
+}
+
+type MatcherResults struct {
+	FileType   scanner.FileType    `json:"filetype,omitempty"`
+	Parameters []scanner.Parameter `json:"parameters,omitempty"`
+	Errors     []MatcherResult     `json:"errors,omitempty"`
+	Infos      []MatcherResult     `json:"infos,omitempty"`
+	Secrets    []MatcherResult     `json:"secrets,omitempty"`
+}
+
+type MatcherResult struct {
+	Name  string `json:"name"`
+	Match string `json:"match"`
+}
+
+func GetJSONString(
```

Comment:
We should add test cases to understand if this function behaves correctly even with edge cases.

For example the filetype match field is always present even if blank:
```
... "matches":{"filetype":{}} ...
```

tested with https://www.edoardoottavianelli.it/

Reply:
I can't find a way to make this behave properly, it seems it happens only for maps and behaves correctly for empty lists ...
As for adding test cases for edge cases, it makes sense.

Reply:
It seems the background of the issue is more complex: empty structs are not counted as 'empty' values by the JSON marshaller. There is an open issue here: https://github.com/golang/go/issues/11939
It does work with empty lists which is why `errors`, `infos`, `secrets`, and parameters are correctly processed as empty.

Reply:
What about https://pkg.go.dev/reflect#DeepEqual? We could just compare the value with an empty filetype struct and put a nil value in the appropriate Matches field if they are equal

Reply:
Empty structs will evaluate as `{}` in the resulting JSON string, that's the issue discussed in the linked GitHub issue. I've added a test case as well, will add more when I get more time.

Reply:
yeah sorry, now i got that